fullstory-segment-middleware/src/main/java/com/fullstorydev/fullstory_segment_middleware/FullStorySegmentMiddleware.java

```diff
@@ -159,4 +167,118 @@ private BasePayload getNewPayloadWithFSURL(BasePayload payload, Map<String, Obje
         }
         return null;
     }
+
+    private Map<String, Object> getSuffixedProps (Map<?, ?> props) {
+        // transform props to comply with FS custom events requirement
+        // TODO: Segment should fail with circular dependency, but we should check anyway
+        Map<String, Object> mutableProps = new HashMap<>();
+        Stack<Map<?,?>> stack = new Stack<>();
+
+        stack.push(props);
+
+        while (!stack.empty()) {
+            Map<?,?> map = stack.pop();
+            for (Map.Entry<?, ?> entry : map.entrySet()) {
+                // We should only be getting String type keys, but parse the map to make sure the keys we use are all String
+                String key = String.valueOf(entry.getKey());
+                Object item = entry.getValue();
+
+                if (item instanceof Map) {
+                    // nested maps, concatenate keys and push back to stack
+                    Map<?,?> itemMap = (Map<?,?>) item;
+                    for (Map.Entry<?, ?> e : itemMap.entrySet()) {
+                        String concatenatedKey =  key + '.' + e.getKey();
```

Comment:
It looks like you're building the selector path for nested objects here. I think it would be a good idea to add a clarifying comment.